rfcs/01924_tls_attestation_ca.md

```diff
@@ -0,0 +1,169 @@
+    to AMD)
+
+It is assumed that prior to the attestation **Server** remotely attests its TEE
+platform key with the **Provider**. **Provider** sends back a signature rooted
+in some well-known root key (e.g. AMD root key), thus confirming the
+authenticity of TEE platform keys.
+
+### Attestation Workflow
+
+The detailed workflow of the remote attestation is represented by 2 stages.
+
+First stage corresponds to the backend application attestation, which happens
+only once when the backend application starts:
+
+1. **Server** generates a TLS key pair
+1. **Server** connects to the **CA**
+1. **CA** attests the **Server** using a remote attestation protocol
+   corresponding to the Server’s TEE
+   1. When requesting a TEE report from the TEE firmware, **Server** also puts
+      the hash of its TLS public key into it
+   1. For example, remote attestation could be done via
+      [#1929](https://github.com/project-oak/oak/pull/1929)
+1. **Server** sends **CA** a certificate signing request
+   1. Communication happens over a secure channel creates as part of the remote
+      attestation
+1. **CA** creates a signed certificate and sends it back to the **Server**
+   1. The certificate contains **Server**'s TEE report
+
+<!-- From: -->
+<!-- https://sequencediagram.googleplex.com/view/6235412701904896 -->
+<img src="images/tls_attestation_ca_workflow_server.png" width="850">
+
+Second stage corresponds to the client connection:
+
+1. [_Optional_] **Client** remotely attests the **CA**
+1. **CA** sends **Client** its root certificate
+   1. **Client** trusts that this is a correct certificate, since it was sent
+      via a secure channel created during the attestation process
+   1. **Client** needs to request the root certificate only once at startup, and
+      then the client can use it for all **Server** connections
+1. **Client** connects to the **Server** using TLS
+1. **Server** sends **Client** a TLS certificate previously signed by **CA**
+
+1. **Client** checks the validity of the **Server**’s TLS certificate
+   1. **Client** verifies that the TLS certificate provided by the **Server**
+      was signed by the **CA**’s root certificate
+   1. **Client** also checks the TEE report provided in the **Server**’s TLS
+      certificate
+      1. It checks that the TEE report was generated using the genuine TEE
+         firmware key
+      1. It checks that the TEE report contains the hash of the **Server**’s
+         public key
+   1. If the **Server**’s TLS certificate is not valid, then the **Client**
+      closes the connection and aborts the protocol
+1. Since **Client** trusts **CA** (and has its root certificate) it can now
+   trust **Server** and establish a secure connection
+
+<!-- From: -->
+<!-- https://sequencediagram.googleplex.com/view/6490702202535936 -->
+<img src="images/tls_attestation_ca_workflow_client.png" width="850">
+
+Several important notes:
```

Comment:
This does not address the TOCTOU issue. For the attestation to have any meaning, the server must be incapable of changing any executable code in its memory while still retaining access to private key.

Reply:
As long as the server is using code/data from the original attestation hash, self-modification by the server code should still be allowed, no? (Not that it's necessarily a good idea or useful)

Reply:
I guess it depends on the type of TEE. If it is a full virtual machine with a separate OS, the entire OS must be validated to ensure that there is no way to e.g. stop this process and start a different one after the certificate is received.